README.md
# targaryen

This repository if to bring up a wordpress server along with a monitoring server.

You can use the cloudformation template to bring up all the required resources for the same. The cloudformation will spin up a t2-medium instance and with the help of chef recepies it'll install wordpress + mysql container as well as telgraf + influxdb + grafana containers.

Once cloudformation is complete, please go to the output section and grab the public ip of the instance that was created. In your brower please go to http://<public-ip>:8080 to visit wordpress site and http://<public-ip>:3000 to visit the grafana dashboard with all metrics being monitored there. The first time credentials to login into Grafana UI is admin:admin

It is easy to scale up this solution to separate out the different layers but I haven't done that considering the cost involved.

As an ideal or the next step to this solution would be to separte the mysql container into a RDS instance perhaps, move the influxdb and grafana containers separately and let telegraf run on the other servers where we want to monitor the containers and send the metrics to influxdb(ofcourse we'll need a route53 domain perhaps for this to make sure we update the influxdb url in there and let all the telgraf containers point to that entry using chef.) Wer can also use EFS if we want to have HA for our wordpress servers.


common/CHANGELOG.md
# common CHANGELOG

This file is used to list changes made in each version of the common cookbook.

## 0.1.0
- [your_name] - Initial release of common

- - -
Check the [Markdown Syntax Guide](http://daringfireball.net/projects/markdown/syntax) for help with Markdown.

The [Github Flavored Markdown page](http://github.github.com/github-flavored-markdown/) describes the differences between markdown on github and standard markdown.


common/recipes/wordpress_deploy.rb
#
# Cookbook Name:: common
# Recipe:: default
#
# Copyright 2018, YOUR_COMPANY_NAME
#
# All rights reserved - Do Not Redistribute
#
common_app = search("aws_opsworks_app", "name:common").first
mysql_root_pass = common_app[:environment][:DBPassword]
mysql_username = common_app[:environment][:DBUsername]
mysql_password = common_app[:environment][:DBPassword]


docker_service 'default' do
  action [:create, :start]
end

docker_image 'wordpress' do
  tag "#{node['wordpress']['release_version']}"
  action :pull
end

docker_image 'mysql' do
  tag "#{node['mysql']['release_version']}"
  action :pull
end

create_dir("#{node['common']['install_path']}/mysql/datadir",0750)
create_dir("#{node['common']['install_path']}/wordpress/wp-content",0750)

docker_container "mysql-wordpress" do
  repo "mysql"
  tag "#{node['mysql']['release_version']}"
  volumes ["#{node['common']['install_path']}/mysql/datadir:/var/lib/mysql" ]
  env ["MYSQL_ROOT_PASSWORD=#{mysql_root_pass}", "MYSQL_DATABASE=wordpress", "MYSQL_USER=#{mysql_username}", "MYSQL_PASSWORD=#{mysql_password}" ]
  restart_policy 'always'
  action :run
end

docker_container "wordpress-app" do
  repo "wordpress"
  tag "#{node['wordpress']['release_version']}"
  volumes ["#{node['common']['install_path']}/wordpress/wordpress-app/wp-content:/var/www/html/wp-content" ]
  env ["WORDPRESS_DB_USER=#{mysql_username}", "WORDPRESS_DB_PASSWORD=#{mysql_password}" ]
  links ['mysql-wordpress:mysql']
  port '8080:80'
  restart_policy 'always'
  action :run
end


common/recipes/monitoring_agent_telegraf_deploy.rb
#
# Cookbook Name:: common
# Recipe:: default
#
# Copyright 2018, YOUR_COMPANY_NAME
#
# All rights reserved - Do Not Redistribute
#
instance = search("aws_opsworks_instance", "self:true").first
hostname = instance[:hostname].dup
pub_ip = instance[:public_ip]
pri_ip = instance[:private_ip]
influxdb_url = "http://#{pub_ip}:8086"
common_app = search("aws_opsworks_app", "name:common").first
grafana_admin_password = common_app[:environment][:GrafanaPassword]
grafana_admin_username = common_app[:environment][:GrafanaUsername]


docker_service 'default' do
  action [:create, :start]
end

docker_image 'telegraf' do
  tag "#{node['telegraf']['release_version']}"
  action :pull
end

create_dir("#{node['common']['install_path']}/telegraf",0750)

template "#{node['common']['install_path']}/telegraf/telegraf.conf" do
  source 'telegraf.conf.erb'
  variables({
    :influxdb_url => "#{influxdb_url}"
  })
  action :create
  owner 'root'
  group 'root'
  mode 0664
end

docker_container "telegraf" do
  repo "telegraf"
  tag "#{node['telegraf']['release_version']}"
  volumes ["#{node['common']['install_path']}/telegraf/telegraf.conf:/etc/telegraf/telegraf.conf:ro", "/var/run/docker.sock:/var/run/docker.sock" ]
  port '8094:8094'
  host_name "#{hostname}"
  restart_policy 'always'
  action :run
end

bash "Grafana init" do
    code <<-EOH
    curl --user #{grafana_admin_username}:#{grafana_admin_password} 'http://#{pri_ip}:3000/api/datasources' -X POST -H 'Content-Type: application/json;charset=UTF-8' --data-binary '{"name":"telegraf","isDefault":true ,"type":"influxdb","url":"http://#{pri_ip}:8086","database":"telegraf","access":"proxy","basicAuth":false}'

    curl --user #{grafana_admin_username}:#{grafana_admin_password} 'http://#{pri_ip}:3000/api/dashboards/import' -X POST -H 'Content-Type: application/json' -d '@#{node['common']['install_path']}/grafana/docker-metrics-per-container_rev2.json'
    EOH
end


common/recipes/monitoring_dashboard_deploy.rb
#
# Cookbook Name:: common
# Recipe:: default
#
# Copyright 2018, YOUR_COMPANY_NAME
#
# All rights reserved - Do Not Redistribute
#
common_app = search("aws_opsworks_app", "name:common").first
influxdb_admin_username = common_app[:environment][:DBUsername]
influxdb_admin_password = common_app[:environment][:DBPassword]
influxdb_username = common_app[:environment][:DBUsername]
influxdb_password = common_app[:environment][:DBPassword]
grafana_admin_password = common_app[:environment][:GrafanaPassword]
grafana_admin_username = common_app[:environment][:GrafanaUsername]
instance = search("aws_opsworks_instance", "self:true").first
pub_ip = instance[:public_ip]


docker_service 'default' do
  action [:create, :start]
end

docker_image 'influxdb' do
  tag "#{node['influxdb']['release_version']}"
  action :pull
end

docker_image 'grafana/grafana' do
  action :pull
end

create_dir("#{node['common']['install_path']}/influxdb/data",0750)

create_dir("#{node['common']['install_path']}/grafana",0750)

cookbook_file "#{node['common']['install_path']}/influxdb/influxdb.conf" do
    source "default/influxdb.conf"
    mode "0644"
    owner 'root'
    group 'root'
end

cookbook_file "#{node['common']['install_path']}/grafana/docker-metrics-per-container_rev2.json" do
    source "default/docker-metrics-per-container_rev2.json"
    mode "0644"
    owner 'root'
    group 'root'
end


docker_container "influxdb" do
  repo "influxdb"
  tag "#{node['influxdb']['release_version']}"
  volumes ["#{node['common']['install_path']}/influxdb/data:/var/lib/influxdb", "#{node['common']['install_path']}/influxdb/influxdb.conf:/etc/influxdb/influxdb.conf:ro" ]
  env ["INFLUXDB_ADMIN_PASSWORD=#{influxdb_admin_password}", "INFLUXDB_ADMIN_USER=#{influxdb_admin_username}", "INFLUXDB_USER_PASSWORD=#{influxdb_password}", "INFLUXDB_USER=#{influxdb_username}", "INFLUXDB_DB=telegraf", "INFLUXDB_ADMIN_ENABLED=true" ]
  port '8086:8086'
  restart_policy 'always'
  action :run
end

docker_container "grafana" do
  repo "grafana/grafana"
  volumes ["#{node['common']['install_path']}/grafana:/var/lib/grafana" ]
  env ["GF_SECURITY_ADMIN_PASSWORD=#{grafana_admin_password}", "GF_SECURITY_ADMIN_USER=#{grafana_admin_username}"]
  port '3000:3000'
  restart_policy 'always'
  action :run
end

docker_service 'default' do
  action [:restart]
end
